ledger/apply/application_test.go

```diff
@@ -962,6 +963,58 @@ func TestAppCallApplyUpdate(t *testing.T) {
 	a.Equal([]byte{2}, br.AppParams[appIdx].ApprovalProgram)
 	a.Equal([]byte{2}, br.AppParams[appIdx].ClearStateProgram)
 	a.Equal(basics.EvalDelta{}, ad.EvalDelta)
+
+	// check app program len
+	appr := make([]byte, 6050, 6050)
+
+	for i := range appr {
+		appr[i] = 2
+	}
+	appr[0] = 4
+	ac = transactions.ApplicationCallTxnFields{
+		ApplicationID:     appIdx,
+		OnCompletion:      transactions.UpdateApplicationOC,
+		ApprovalProgram:   appr,
+		ClearStateProgram: []byte{2},
+	}
+	params = basics.AppParams{
+		ApprovalProgram: []byte{1},
+		StateSchemas: basics.StateSchemas{
+			GlobalStateSchema: basics.StateSchema{NumUint: 1},
+		},
+		ExtraProgramPages: 1,
+	}
+	h = transactions.Header{
+		Sender: sender,
+	}
+
+	b.balances = make(map[basics.Address]basics.AccountData)
+	cbr = basics.AccountData{
+		AppParams: map[basics.AppIndex]basics.AppParams{appIdx: params},
+	}
+	cp = basics.AccountData{
+		AppParams: map[basics.AppIndex]basics.AppParams{appIdx: params},
+	}
+	b.balances[creator] = cp
+	b.appCreators = map[basics.AppIndex]basics.Address{appIdx: creator}
+
+	// check program len check doesn't happen in current consensus proto version
+	b.SetProto(protocol.ConsensusV28)
+	proto = b.ConsensusParams()
+	ep.Proto = &proto
+
+	b.pass = true
+	err = ApplicationCall(ac, h, &b, ad, &ep, txnCounter)
+	a.NoError(err)
```

Comment:
Why does this work?  You are updating with a 6k program, under the old proto.  That should fail.  Am I misreading?

Reply:
this works because b is a mocked Balances obj and it actually invokes this mocked StatefulEval https://github.com/algorand/go-algorand/blob/af3858cf92829aa8f6eb2ffcedd4a03c0b820c88/ledger/apply/application_test.go#L189. 

since I wanted to test updateApplication() only, b.pass is set to true at line 1006.

Reply:
ApplicationCall should still have failed, I think, even before it got to StatefulEval().  Your extra program pages is 1 in params.  But you are calling this with UpdateApplicationOC and a 6k approval program.  Why does updateApplication allow such a big approval program?

I don't think we can trust these kinds of tests for actual transaction execution very much.  We need to put them in e2e subs tests against a running network or we'll never have the level of confidence we need.

Reply:
I see the bug.  In the new code, you don't check the length at all in old versions.  But you must check the length - you must maintain the same behavior in old versions.  So the behaviour should be:

 Old version: check against a single program page of space
 New version: check against the extra pages specified in params

Reply:
I'll add an e2e test. 
do I need to add a check for the old version in updateApplication() though? 
doesn't these checks, https://github.com/algorand/go-algorand/blob/e3e419754c198b5232927c735763b36390e47d68/data/transactions/transaction.go#L405-L414 enforce that the old version would maintain the same behavior?

Reply:
I think you are right.  I find it a bit confusing because in old code, a too big program is rejected in WellFormed, while in the new code, it's rejected in updateApplication.  When just reading updateApplication, it appears we have no check at all unless `proto.EnableExtraPagesOnAppUpdate` is true.

Could you add a comment in updateApplication explaining that WellFormed rejects all multipage program updates when `proto.EnableExtraPagesOnAppUpdate` is false? Or else simply reject them here too, even though it's redundant.